From 9fcf5eb18dc93862f73b69199226def74b91586c Mon Sep 17 00:00:00 2001
From: Hauke Mehrtens <hauke.mehrtens@intel.com>
Date: Thu, 19 Nov 2020 18:31:14 +0100
Subject: [PATCH] PONRTSYS-8285: Move the deactivating of multicast filter to
 later place

Instead of checking in the beginning of the filter if the multicast packages
should get filtered to this in a later place to allow other stuff to happen
in addition.
---
 drivers/net/datapath/mcast_bridge_filter/mcast_bridge_filter.c | 8 ++++----
 1 file changed, 4 insertions(+), 4 deletions(-)

diff --git a/drivers/net/datapath/mcast_bridge_filter/mcast_bridge_filter.c b/drivers/net/datapath/mcast_bridge_filter/mcast_bridge_filter.c
index f5a6682245af..e392a101b6d6 100644
--- a/drivers/net/datapath/mcast_bridge_filter/mcast_bridge_filter.c
+++ b/drivers/net/datapath/mcast_bridge_filter/mcast_bridge_filter.c
@@ -56,6 +56,9 @@ static unsigned int dp_mcast_br_hook_ipv4(struct sk_buff *skb, int nhoff)
 {
 	struct iphdr *iph;
 
+	if (!dp_mcast_fliter_active)
+		return NF_ACCEPT;
+
 	if (!pskb_network_may_pull(skb, nhoff + sizeof(struct iphdr)))
 		return NF_ACCEPT;
 
@@ -93,7 +96,7 @@ static unsigned int dp_mcast_br_hook_ipv6(struct sk_buff *skb, int nhoff)
 		return NF_ACCEPT;
 	}
 
-	if (ipv6_is_mld(skb, nexthdr, nhoff)) {
+	if (ipv6_is_mld(skb, nexthdr, nhoff) && dp_mcast_fliter_active) {
 		pr_debug("IPv6 MLD packet dropped");
 		return NF_DROP;
 	}
@@ -107,9 +110,6 @@ static unsigned int dp_mcast_br_hook(void *priv, struct sk_buff *skb,
 	int nhoff = skb_network_offset(skb);
 	int err;
 
-	if (!dp_mcast_fliter_active)
-		return NF_ACCEPT;
-
 	if (eth_type_vlan(proto)) {
 		err = dp_mcast_vlan_iter(skb, &proto, &nhoff);
 		if (err) {
